### PR TITLE
feat(execution): implement a versioned cell data structure for implem…

### PR DESCRIPTION
**Pull request by MohammadNassar1**
…enting concurrency

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/1420)

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/1420)* status: all files reviewed, 7 unresolved discussions (waiting on @giladchase and @MohammadNassar1)

___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 7 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NpozWMw5ycOPEG1dO3X:-NpozWMw5ycOPEG1dO3Y:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L7)):*
> ```XML
> pub mod test;
> 
> pub(crate) type VersionId = u64;
> ```





_Suggestion:_
```XML
Version
```
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 10 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NpoywLRDNRbi4S8QQNu:-NpoywLRDNRbi4S8QQNv:b-93r5aa) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L10)):*
> ```XML
> 
> #[derive(Clone)]
> pub struct VersionedCell<T> {
> ```

Please add a succinct docstring.



_Code quote:_
```XML
VersionedCell
```
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 12 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-Npoz1KN5ye9a8nkhXD7:-Npoz1KN5ye9a8nkhXD8:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L12)):*
> ```XML
> pub struct VersionedCell<T> {
>     pub initial_value: T,
>     pub write_versions: BTreeMap<VersionId, T>,
> ```





_Suggestion:_
```XML
pub writes: BTreeMap<VersionId, T>,
```
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 20 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-Npoz5tc0zX2pGtQUFg4:-Npoz5tc0zX2pGtQUFg5:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L20)):*
> ```XML
>     }
> 
>     // Retrieving the most recent value written by a version less than or equal to the provided
> ```





_Suggestion:_
```XML
Retrieves
```
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 22 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-Npoz_usCGRns05SC7It:-Npoz_usCGRns05SC7Iu:b-6ggzxn) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L22)):*
> ```XML
>     // Retrieving the most recent value written by a version less than or equal to the provided
>     // version.
>     pub fn read(&self, id: VersionId) -> &T {
> ```

You don't think it's a good name?



_Suggestion:_
```XML
version: Version
```
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 24 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NpozoTn8eir289omNRY:-NpozoTn8eir289omNRZ:beq7ngl) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L24)):*
> ```XML
>     pub fn read(&self, id: VersionId) -> &T {
>         let closest_version =
>             self.write_versions.range(..=id).next_back().map(|(&version, _)| version);
> ```

Is this done in `O(n)`?



_Code quote:_
```XML
        let closest_version =
            self.write_versions.range(..=id).next_back().map(|(&version, _)| version);
```
___
*[`crates/blockifier/src/concurrency/versioned_cell_test.rs` line 6 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-Npp-sEG5kthTnlvLojw:-Npp-sEG5kthTnlvLojx:b-iz3pkl) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell_test.rs#L6)):*
> ```XML
> 
> #[test]
> fn test_versioned_cell() {
> ```

Please restore the Pythonic test comments, they are good.



_Code quote:_
```XML
test_versioned_cell
```

**Review by giladchase — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/1420)* status: all files reviewed, 8 unresolved discussions (waiting on @MohammadNassar1)

___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 27 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NpsfuVD8--cxHKkwSsn:-NpsfuVD8--cxHKkwSso:bs819m8) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L27)):*
> ```XML
> 
>         let last_value = closest_version.and_then(|version| self.write_versions.get(&version));
>         if let Some(last_value) = last_value { last_value } else { &self.initial_value }
> ```

Or alternatively `map(|(_, value) value).unwrap_or(&self.initial_value)` at the end.



_Suggestion:_
```XML
    self.write_versions
        .range(..=id)
        .next_back()
        .map_or(&self.initial_value, |(_, value)| value)
```

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/1420)* status: 3 of 4 files reviewed, 5 unresolved discussions (waiting on @elintul and @giladchase)

___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 10 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NpoywLRDNRbi4S8QQNu:-NpthVRGDnN7tlYaOfkO:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L10)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Please add a succinct docstring.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 22 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-Npoz_usCGRns05SC7It:-Nptg94C98H-GZIyF0bO:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L22)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

You don't think it's a good name?
</blockquote></details>

Done.
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 24 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NpozoTn8eir289omNRY:-Nptg9dP-wZikRrbT1Ft:bqxawix) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L24)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Is this done in `O(n)`?
</blockquote></details>

No, it's in O(log(n)).
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 27 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NpsfuVD8--cxHKkwSsn:-NptgWkvDr5a9X5Ki8At:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell.rs#L27)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Or alternatively `map(|(_, value) value).unwrap_or(&self.initial_value)` at the end.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/concurrency/versioned_cell_test.rs` line 6 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-Npp-sEG5kthTnlvLojw:-NptiMh08-PqB0nQeznJ:bh3pnyt) ([raw file](https://github.com/starkware-libs/blockifier/blob/45bc1cc6a86fb4d1f8b7da0255470f219e0bbc65/crates/blockifier/src/concurrency/versioned_cell_test.rs#L6)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Please restore the Pythonic test comments, they are good.
</blockquote></details>

The Pythonic test checks the same scenarios as mine.
In addition, it contains dependency graph tests that are not relevant to our implementation.

**Review by elintul — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/1420)* status: all files reviewed, 4 unresolved discussions (waiting on @giladchase and @MohammadNassar1)

___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 13 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-Npu23WH5vg_Af7EDKid:-Npu23WH5vg_Af7EDKie:b-suv37h) ([raw file](https://github.com/starkware-libs/blockifier/blob/33dd0ff19b9921cf3d3cf268787234d09d368645/crates/blockifier/src/concurrency/versioned_cell.rs#L13)):*
> ```XML
> /// different versions of executions.
> /// This allows maintaining the cell with the correct value in the context of each execution.
> #[derive(Clone, Debug, PartialEq, Eq)]
> ```

Alphabetize?



_Code quote:_
```XML
Clone, Debug, PartialEq, Eq
```

**Review by giladchase — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/1420)* status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @elintul and @MohammadNassar1)

___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 30 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NqLwCKU1nN7YEobpEbB:-NqLwCKU1nN7YEobpEbC:b-vkslnd) ([raw file](https://github.com/starkware-libs/blockifier/blob/81c591dae7e55d688afa23d4fc5c85f4aa4abdbf/crates/blockifier/src/concurrency/versioned_cell.rs#L30)):*
> ```XML
>     }
> 
>     pub fn write(&mut self, version: Version, value: T) {
> ```

Will this component only be used synchronously? Or will the reads/writes need to be threadsafe?
___
*[`crates/blockifier/src/concurrency/versioned_cell.rs` line 32 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-NqLuzN81ROVlj1bIeIs:-NqLuzN81ROVlj1bIeIt:bhigrqy) ([raw file](https://github.com/starkware-libs/blockifier/blob/81c591dae7e55d688afa23d4fc5c85f4aa4abdbf/crates/blockifier/src/concurrency/versioned_cell.rs#L32)):*
> ```XML
>     pub fn write(&mut self, version: Version, value: T) {
>         self.writes.insert(version, value);
>     }
> ```

Do we want to allow overwriting the same version with a different value? 

In other words, should it be only possibly to `write` in a fresh version?



_Code quote:_
```XML
    pub fn write(&mut self, version: Version, value: T) {
        self.writes.insert(version, value);
    }
```

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/1420)* status: 3 of 4 files reviewed, 4 unresolved discussions (waiting on @elintul)

*[a discussion](https://reviewable.io/reviews/starkware-libs/blockifier/1420#-Nqvg49AB-ARFSxfGsDq:-Nqvg49AB-ARFSxfGsDr:bvyndgs) (no related file):*
Following our discussion, we have decided to implement the `VersionedState` DS using a two-tier hierarchy:
1.  `VersionedState`
2.  `VersionedStorage`

The `VersionedCell` has been moved to become a part of the `VersionedStorage` in [this](https://reviewable.io/reviews/starkware-libs/blockifier/1421) PR.

___

**Comment by MohammadNassar1** [comment]
Following our discussion, we have decided to implement the `VersionedState` DS using a two-tier hierarchy:
1.  `VersionedState`
2.  `VersionedStorage`

The `VersionedCell` has been moved to become a part of the `VersionedStorage` in [this](https://reviewable.io/reviews/starkware-libs/blockifier/1421) PR.
